File: gam_TMB/gam_tmb_tp.R
library(mgcv)
library(TMB)
library(ggplot2)

set.seed(1)

#Make 20000 sampling locations
N <- 20000
E <- runif(N,-10,10)
N <- runif(N,-10,10)
dat <- data.frame(y=0,E,N) #Fill in y with zeros for now

s0 <- gam(y~s(E,N,bs='ts',k=60),data=dat,fit=FALSE) #Create model matrix
modMat <- s0$X[,-1] #Model matrix (no intercept)
penMat <- s0$smooth[[1]]$S[[1]] #Penalty matrix
penMat <- as(penMat,'sparseMatrix') #Change to sparse matrix
penDim <- nrow(penMat) #Dimensions of penalty matrix

b0 <- 1 #Intercept
(smoothCoefs <- rnorm(ncol(modMat))) #Smoothing coefficients
sigma <- 2 #Residual SD
yhat <- b0 + modMat %*% smoothCoefs #Mean values at each location
dat$y <- rnorm(N,yhat,sigma) #Generate data
dat$field <- modMat %*% smoothCoefs #Spatial field

ggplot(dat)+geom_point(aes(x=E,y=N,col=field)) #Plot spatial field (no noise)
ggplot(dat)+geom_point(aes(x=E,y=N,col=y)) #Plot data

#Input for model
datList <- list(y=dat$y, smoothMat=modMat, penaltyMat=penMat, penaltyDim=penDim) #Data
parList <- list(b0=0, smoothCoefs=rep(0,ncol(modMat)), log_lambda=0, log_sigma=0) #Parameters

compile("gam_tmb_tp.cpp")
dyn.load(dynlib("gam_tmb_tp"))
obj <- MakeADFun(data = datList, parameters = parList, random=c('smoothCoefs'), DLL = "gam_tmb_tp")

#Memory usage spikes from 4.5GB to 16GB during Optimizing Tape stage
opt <- nlminb(obj$par,obj$fn,obj$gr) #Run model
report <- sdreport(obj)

opt$convergence==0 #Did model converge?
report #Looks OK. b0 and sigma are very close to actual values (1 and 2)

#Estimates smoothing coefficients fairly well
ggplot(data.frame(predicted=report$par.random,actual=smoothCoefs))+
  geom_point(aes(x=predicted,y=actual))+geom_abline(intercept = 0,slope=1)+
  labs(title='Smoothing coefficients')

File: gam_TMB/gam_tmb_tp.cpp
#include <TMB.hpp>  
template<class Type>
Type objective_function<Type>::operator() (){
  using namespace density;
  using namespace Eigen; 
  
  //2D thin-plate regression spline from mgcv
  
  //Read data from R------------------
  DATA_VECTOR(y);                     // Response
  DATA_MATRIX(smoothMat);             // Design matrix (no intercept)
  DATA_SPARSE_MATRIX(penaltyMat);     // Penalty matrix
  DATA_INTEGER(penaltyDim);           // Dimensions of penalty matrix
  
  //Read parameters from R------------
  PARAMETER(b0);                      // Intercept
  PARAMETER_VECTOR(smoothCoefs);      // Smooth coefficients
  PARAMETER(log_lambda);              // Log penalization parameter
  PARAMETER(log_sigma);               // Log(SD) of measurement error 
  
  //Transform SD and penalization parameters 
  Type sigma = exp(log_sigma);
  Type lambda = exp(log_lambda);
  
  //Initialize objective function
  Type nll=0;
  
  // Penalization ------------------------------------
  
  SparseMatrix<Type> S = lambda*penaltyMat; //Penalty term * Penalty matrix
  nll -= 0.5*penaltyDim*log_lambda - 0.5*GMRF(S).Quadform(smoothCoefs);
  
  // Main model -------------------------------------------
  vector<Type> mu = b0 + smoothMat*smoothCoefs; //Expected value
  
  nll -= sum(dnorm(y, mu, sigma, true)); //Decrement logLik
  
  return nll; //Return likelihood
}
